### PR TITLE
Multiple namespaces support

### PR DESCRIPTION
**Pull request by melkio**
Connects to #368

### CODE REVIEW

File: Snippets/ASB/ASB_7/MultipleNamespaces.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    using NServiceBus;
+    using NServiceBus.AzureServiceBus.Addressing;
+
+    class MultipleNamespaces
+    {
+        private readonly EndpointConfiguration endpointConfiguration;
+
+        public MultipleNamespaces(EndpointConfiguration endpointConfiguration)
+        {
+            this.endpointConfiguration = endpointConfiguration;
```
Comment (SimonCropp) [inline Snippets/ASB/ASB_7/MultipleNamespaces.cs]: snippets dont need to be proper code. just pass in an `EndpointConfiguration endpointConfiguration` to each method if u need an insatnce

Comment (SimonCropp) [inline Snippets/ASB/ASB_7/MultipleNamespaces.cs]: or just use the null field

File: nservicebus/azure-servicebus/multiple-namespaces-support.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+---
+title: Multiple namespace support
+summary: How to configure transport to support different partitioning strategies
```
Comment (SimonCropp) [inline nservicebus/azure-servicebus/multiple-namespaces-support.md]: summary should be proper grammar. so ending in a `.`

Comment (SimonCropp) [inline nservicebus/azure-servicebus/multiple-namespaces-support.md]: perhaps 

> Configuring Azure ServiceBus transport to support different partitioning strategies.

File: Snippets/ASB/ASB_7/MultipleNamespaces.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    using NServiceBus.AzureServiceBus.Addressing;
+
+    class MultipleNamespaces
+    {
+        public void SingleNamespaceStrategyWithAddNamespace(EndpointConfiguration endpointConfiguration)
+        {
+            #region single_namespace_partitioning_strategy_with_add_namespace
+
+            endpointConfiguration.UseTransport<AzureServiceBusTransport>()
+                .NamespacePartitioning().UseStrategy<SingleNamespacePartitioning>()
```
Comment (SeanFeldman) [inline Snippets/ASB/ASB_7/MultipleNamespaces.cs]: Would it make sense to have the on 2 separate lines?

```c#
   .NamespacePartitioning()
   .UseStrategy<SingleNamespacePartitioning>()
```

File: Snippets/ASB/ASB_7/MultipleNamespaces.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            #region sharded_partitioning_strategy_configuration
+
+            endpointConfiguration.UseTransport<AzureServiceBusTransport>()
+                .NamespacePartitioning().UseStrategy<ShardedNamespacePartitioning>()
+                    .AddNamespace("namespace1", "Endpoint=sb://namespace1.servicebus.windows.net;SharedAccessKeyName=[shared access key name];SharedAccessKey=[shared access key]")
+                    .AddNamespace("namespace2", "Endpoint=sb://namespace2.servicebus.windows.net;SharedAccessKeyName=[shared access key name];SharedAccessKey=[shared access key]")
+                    .AddNamespace("namespace3", "Endpoint=sb://namespace3.servicebus.windows.net;SharedAccessKeyName=[shared access key name];SharedAccessKey=[shared access key]");
+
+            #endregion
```
Comment (SeanFeldman) [inline Snippets/ASB/ASB_7/MultipleNamespaces.cs]: Should `ShardedNamespacePartitioning` cover registration of the sharding rules? 
As a reader I would not know how `ShardedNamespacePartitioning` is different from `ReplicatedNamespacePartitioning` or `RoundRobinNamespacePartitioning`

File: nservicebus/azure-servicebus/multiple-namespaces-support.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+summary: Configuring Azure ServiceBus transport to support different partitioning strategies. 
+component: ASB
+tags:
+- Cloud
+- Azure
+- Transports 
+reviewed: 2016-04-18
+---
+
+// INTRODUCTION HERE!!!
```
Comment (SeanFeldman) [inline nservicebus/azure-servicebus/multiple-namespaces-support.md]: Introduction NOW! :smile:

**Comment by melkio** [comment]
@Particular/azure-service-bus-maintainers ping

File: nservicebus/azure-servicebus/multiple-namespaces-support.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+To configure _namespace partitioning strategy_ properly users have to select desired `INamespacePartitioningStrategy` implementation and map related namespaces, as showed in the next snippets, for each supported strategies.
```
Comment (SeanFeldman) [inline nservicebus/azure-servicebus/multiple-namespaces-support.md]: >To configure _namespace partitioning strategy_ 

Aren't we configuring namespace partitioning _by specifying_ a strategy?

File: Snippets/ASB/ASB_7/MultipleNamespaces.cs (hunk 4 above)
Comment (melkio) [inline Snippets/ASB/ASB_7/MultipleNamespaces.cs]: There will be a detailed description inside the md file to describe the differences between strategies. Snippets are only to show how to configure a specific strategy

File: nservicebus/azure-servicebus/multiple-namespaces-support.md (hunk 6 above)
Comment (SeanFeldman) [inline nservicebus/azure-servicebus/multiple-namespaces-support.md]: > users have to select 

I can't recall doco talking about users. Perhaps this would be easier to change if we first have explanation (intro) about what Namespace Partitioning is.

File: nservicebus/azure-servicebus/multiple-namespaces-support.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+### Single namespace partitioning #
+
+The default _namespace partitioning strategy_ is `SingleNamespacePartitioning`. With the default setup, transports uses only a single namespace to send and receive messages.
```
Comment (SeanFeldman) [inline nservicebus/azure-servicebus/multiple-namespaces-support.md]: > transports

Singular?

File: nservicebus/azure-servicebus/multiple-namespaces-support.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+---
+
+// INTRODUCTION HERE!!!
+
+To configure _namespace partitioning strategy_ properly users have to select desired `INamespacePartitioningStrategy` implementation and map related namespaces, as showed in the next snippets, for each supported strategies.
+
+### Single namespace partitioning #
+
+The default _namespace partitioning strategy_ is `SingleNamespacePartitioning`. With the default setup, transports uses only a single namespace to send and receive messages.  
+Users can provide namespace connection string directly with `ConnectionString` extension method:
```
Comment (SeanFeldman) [inline nservicebus/azure-servicebus/multiple-namespaces-support.md]: > Users...
What if they are customers? :)

An option:

Namespace connection string can be defined via code using `ConnectionString` extension:

File: nservicebus/azure-servicebus/multiple-namespaces-support.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+- Azure
+- Transports 
+reviewed: 2016-04-18
+---
+
+// INTRODUCTION HERE!!!
+
+To configure _namespace partitioning strategy_ properly users have to select desired `INamespacePartitioningStrategy` implementation and map related namespaces, as showed in the next snippets, for each supported strategies.
+
+### Single namespace partitioning #
```
Comment (SimonCropp) [inline nservicebus/azure-servicebus/multiple-namespaces-support.md]: why the trailing `#`

File: nservicebus/azure-servicebus/multiple-namespaces-support.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+or selecting explicitly the `SingleNamespacePartitioning` strategy and providing a map between namespace name and namespace connection string.  
+Transport uses namespace name to avoid sharing namespace secrets.
+
+snippet: single_namespace_partitioning_strategy_with_add_namespace
+
+With this strategy, only one namespace has to be configured. Configuration throws a `ConfigurationErrorsException` if more than one namespace has been configured.   
+`SingleNamespacePartitioning` strategy is the easiest one but it's the less relyable: the only configured namespace is a single point of failure for the entire system.
+
+### Round robin namespace partitioning #
```
Comment (SimonCropp) [inline nservicebus/azure-servicebus/multiple-namespaces-support.md]: trialing hash

**Comment by melkio** [comment]
@Particular/azure-service-bus-maintainers @weralabaj almost ready

Few notes:
- we have to decide which strategies will be delivered and then fix this article accordingly
- article doesn't  cover `SetShardingRule` method, because of this: https://github.com/Particular/NServiceBus.AzureServiceBus/issues/189
- does upgrade guide need a quick description of _multiple namespace support_? It's a new feature. No need to upgrade something

**Comment by SeanFeldman** [comment]
@melkio upgrade guide should focus on things that are important when upgrading an existing code. New features should be covered by the regular doco. In case upgrade is not possible w/o using new features, then those should be mentioned briefly and linked to the main doco.

**Comment by SeanFeldman** [comment]
> @Particular/azure-service-bus-maintainers @weralabaj almost ready

Let's agree to ping when it's ready? :)
We shouldn't comment on things that are still a wip. Makes author life hard.

**Comment by melkio** [comment]
@SeanFeldman from my point of view is completed: I covered and described all strategies and related scenarios. I said `almost` because we have to decide if we want to deliver all the strategies or move some of them as samples, as discussed with @yvesgoeleven this afternoon.

If you want you can review IMO. We can decide later what strategies will be confirmed and what will not

**Comment by SeanFeldman** [comment]
@melkio Thank you for explanation.
The point I was trying to make is that as a maintainer that gets pinged with message "almost ready" it's confusing

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+summary: Configuring Azure ServiceBus transport to support different partitioning strategies. 
+component: ASB
+tags:
+- Cloud
+- Azure
+- Transports 
+reviewed: 2016-04-18
+---
+
+`Azure Service Bus` transport enables different _namespace partitioning strategies_ to cover different scenarios like High Availability and multiple Data Center support or to overtake [service limits](https://azure.microsoft.com/en-us/documentation/articles/service-bus-quotas/).
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > transport enables different _namespace partitioning strategies_ to cover different scenarios like

Perhaps:
"transport supports different _namespace partitioning strategies_ to cover different scenarios such as"

Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > or to overtake

"to overcome" ?

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+tags:
+- Cloud
+- Azure
+- Transports 
+reviewed: 2016-04-18
+---
+
+`Azure Service Bus` transport enables different _namespace partitioning strategies_ to cover different scenarios such as High Availability and multiple Data Center support or to overcome [service limits](https://azure.microsoft.com/en-us/documentation/articles/service-bus-quotas/).  
+
+Configuration APIs give the possibility to select desired _namespace partitioning strategies_ and then to configure specific namespace(s) to use.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: If I read correctly, this is the intent?

"_Namespace partitioning strategy_ can be configured using Azure Service Bus configuration API, specifying namespace(s) to use."

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+## Single namespace partitioning ##
+
+The default _namespace partitioning strategy_ is `SingleNamespacePartitioning`. With the default setup, transport uses only a single namespace to send and receive messages.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: >  With the default setup, 

Not sure what it means. 
The rest of the sentence explains what it does.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+- Azure
+- Transports 
+reviewed: 2016-04-18
+---
+
+`Azure Service Bus` transport enables different _namespace partitioning strategies_ to cover different scenarios such as High Availability and multiple Data Center support or to overcome [service limits](https://azure.microsoft.com/en-us/documentation/articles/service-bus-quotas/).  
+
+Configuration APIs give the possibility to select desired _namespace partitioning strategies_ and then to configure specific namespace(s) to use.
+
+## Single namespace partitioning ##
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: Double hash at the end

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+snippet: single_namespace_partitioning_strategy_with_default_connection_string
+
+or selecting explicitly the `SingleNamespacePartitioning` strategy and providing a map between namespace name and namespace connection string.  
+Transport uses namespace name to avoid sharing namespace secrets.
+
+snippet: single_namespace_partitioning_strategy_with_add_namespace
+
+With this strategy, only one namespace has to be configured. Configuration throws a `ConfigurationErrorsException` if more than one namespace has been configured.   
+`SingleNamespacePartitioning` strategy is the easiest one but it's the less relyable: the only configured namespace is a single point of failure for the entire system.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > it's the less relyable

Not sure if reliability is the point here.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+## Round robin namespace partitioning ##
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: double hash at the end

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+Another possible strategy is `RoundRobinNamespacePartitioning`. With this strategy, transport uses each registered namespace cyclically to send messages and it receives messages from all of them.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > Another possible strategy is 

Redundant. From the reader perspective, all the strategies are possible or else would not be listed in the doco :)

Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > With this strategy, transport uses each registered namespace cyclically to send messages and it receives messages from all of them.  

Perhaps like this?

Multiple namespace are used. Messages are sent to a single namespace and received from all the namespace. For sending operations, namespaces are cycled through.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+snippet: single_namespace_partitioning_strategy_with_add_namespace
+
+With this strategy, only one namespace has to be configured. Configuration throws a `ConfigurationErrorsException` if more than one namespace has been configured.   
+`SingleNamespacePartitioning` strategy is the easiest one but it's the less relyable: the only configured namespace is a single point of failure for the entire system.
+
+## Round robin namespace partitioning ##
+
+Another possible strategy is `RoundRobinNamespacePartitioning`. With this strategy, transport uses each registered namespace cyclically to send messages and it receives messages from all of them.  
+This strategy is useful to avoid throttling by Azure infrastructure for capacity limit.
+`RoundRobinNamespacePartitioning` strategy needs more than one configured namespace. Configuration throws a `ConfigurationErrorsException` if only one namespace has been configured.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > `RoundRobinNamespacePartitioning` strategy needs more than one configured namespace. 

Should this come prior?

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+## Round robin namespace partitioning ##
+
+Another possible strategy is `RoundRobinNamespacePartitioning`. With this strategy, transport uses each registered namespace cyclically to send messages and it receives messages from all of them.  
+This strategy is useful to avoid throttling by Azure infrastructure for capacity limit.
+`RoundRobinNamespacePartitioning` strategy needs more than one configured namespace. Configuration throws a `ConfigurationErrorsException` if only one namespace has been configured. 
+
+snippet: round_robin_partitioning_strategy
+
+## Fail over namespace partitioning ##
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: double hash at the end

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+`FailOverNamespacePartitioning` strategy covers High Availability scenario: configuring two different namespaces, a primary and a secondary, transport uses the first one to read or send messages if it's reachable. Otherwise transport switches to secondary.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: :+1: Slight re-phrasing: 

`FailOverNamespacePartitioning` strategy provides transport High Availability by using two different namespaces, a primary and a secondary. The transport uses the primary namespace. When the primary namespace is not available, the transport will switch to the secondary namespace.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+`FailOverNamespacePartitioning` strategy needs exactly two namespaces. Configuration throws a `ConfigurationErrorsException` if zero, one or more than two namespace has been configured.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > if zero, one or more than two namespace has been configured. 

if number of namespaces is different than two.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+
+## Fail over namespace partitioning ##
+
+`FailOverNamespacePartitioning` strategy covers High Availability scenario: configuring two different namespaces, a primary and a secondary, transport uses the first one to read or send messages if it's reachable. Otherwise transport switches to secondary.
+`FailOverNamespacePartitioning` strategy needs exactly two namespaces. Configuration throws a `ConfigurationErrorsException` if zero, one or more than two namespace has been configured. 
+
+snippet: fail_over_partitioning_strategy 
+
+
+## Replicated namespace partitioning ##
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: double hash at the end

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+
+Scenario covered by `RepliactedNamespacePartitioning` is to support publish messages across different data centers. With this strategy, transport uses all configured namespaces to send and receive messages.  
+When transport has to publish a message, it publishes the same message to all namespaces.
+`RepliactedNamespacePartitioning` strategy needs more than one configured namespace. Configuration throws a `ConfigurationErrorsException` if only one namespace has been configured. 
+
+snippet: replicated_partitioning_strategy
+
+## Sharded namespace partitioning ##
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: double hash at the end

Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: Are we going to move this out of the transport and make into a sample?

Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > is to support publish messages across different data centers.

Technically, replication happens on different namespace. Namespaces can be in different data centres.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md (hunk 22 above)
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: Are we going to move this out of the transport and make into a sample?

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+
+`FailOverNamespacePartitioning` strategy covers High Availability scenario: configuring two different namespaces, a primary and a secondary, transport uses the first one to read or send messages if it's reachable. Otherwise transport switches to secondary.
+`FailOverNamespacePartitioning` strategy needs exactly two namespaces. Configuration throws a `ConfigurationErrorsException` if zero, one or more than two namespace has been configured. 
+
+snippet: fail_over_partitioning_strategy 
+
+
+## Replicated namespace partitioning ##
+
+Scenario covered by `RepliactedNamespacePartitioning` is to support publish messages across different data centers. With this strategy, transport uses all configured namespaces to send and receive messages.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > is to support publish messages across different data centers.

Technically, replication happens on different namespace. Namespaces can be in different data centres.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+`FailOverNamespacePartitioning` strategy covers High Availability scenario: configuring two different namespaces, a primary and a secondary, transport uses the first one to read or send messages if it's reachable. Otherwise transport switches to secondary.
+`FailOverNamespacePartitioning` strategy needs exactly two namespaces. Configuration throws a `ConfigurationErrorsException` if zero, one or more than two namespace has been configured. 
+
+snippet: fail_over_partitioning_strategy 
+
+
+## Replicated namespace partitioning ##
+
+Scenario covered by `RepliactedNamespacePartitioning` is to support publish messages across different data centers. With this strategy, transport uses all configured namespaces to send and receive messages.  
+When transport has to publish a message, it publishes the same message to all namespaces.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > publish 

Publish entails pub/sub. Should "Send" be used instead?

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+`RepliactedNamespacePartitioning` strategy needs more than one configured namespace. Configuration throws a `ConfigurationErrorsException` if only one namespace has been configured.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > needs 

requires

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+
+Scenario covered by `RepliactedNamespacePartitioning` is to support publish messages across different data centers. With this strategy, transport uses all configured namespaces to send and receive messages.  
+When transport has to publish a message, it publishes the same message to all namespaces.
+`RepliactedNamespacePartitioning` strategy needs more than one configured namespace. Configuration throws a `ConfigurationErrorsException` if only one namespace has been configured. 
+
+snippet: replicated_partitioning_strategy
+
+## Sharded namespace partitioning ##
+
+`ShardedNamespacePartitioning` is a specification of `RoundRobinNamespacePartitioning`: transport doesn't decide where to publish messages but it's up to user define a `ShardingKey` to route messages to a namespace or to another before publishing a message.
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > specification 

Do you mean it's similar in a way?

Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > transport doesn't decide where to publish messages but it's up to user define a `ShardingKey` to route messages to a namespace or to another before publishing a message.

Messages are sent to the namespace selected by the custom `ShardingKey` defined and provided to the strategy.

Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: There's no code snippet to demonstrate how to supply a `SharingKey` - does that code exist? /cc @yvesgoeleven

**Comment by SeanFeldman** [comment]
@Particular/azure-service-bus-maintainers There was a discussion about getting `RepliactedNamespacePartitioning` and `ShardedNamespacePartitioning` out of the transport and possibly adding as samples. :+1: 

Thanks to @melkio doco it was highlighted that:

* `RepliactedNamespacePartitioning` could have reliability issues that need to be tested
* `ShardedNamespacePartitioning` can't be even configure (separate issue #198)

**Comment by melkio** [comment]
@Particular/azure-service-bus-maintainers +1 for me to move strategies mentioned by @SeanFeldman out of transport. 
I'm not sure with `ReplicatedNamespacePartitioning` and `SharedNamespacePartitioning` implementation. I'd rather to wait before delivering them.

Let me know your opinion. I'll upgrade transport, docu and I'll create samples

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+summary: Configuring Azure ServiceBus transport to support different partitioning strategies. 
+component: ASB
+tags:
+- Cloud
+- Azure
+- Transports 
+reviewed: 2016-04-18
+---
+
+`Azure Service Bus` transport enables different _namespace partitioning strategies_ to cover different scenarios such as High Availability and multiple Data Center support or to overcome [service limits](https://azure.microsoft.com/en-us/documentation/articles/service-bus-quotas/).
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > `Azure Service Bus` 

no need for ``, just plain Azure Service Bus

Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > to cover different scenarios

to cover various scenarios

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+- Azure
+- Transports 
+reviewed: 2016-04-18
+---
+
+`Azure Service Bus` transport enables different _namespace partitioning strategies_ to cover different scenarios such as High Availability and multiple Data Center support or to overcome [service limits](https://azure.microsoft.com/en-us/documentation/articles/service-bus-quotas/).  
+
+_Namespace partitioning strategies_can be configured using Azure Service Bus configuration API, specifying namespace(s) to use.
+
+## Single namespace partitioning #
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: Remove all the hashes after section names

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+
+The default _namespace partitioning strategy_ is `SingleNamespacePartitioning`. With this strategy, transport uses only a single namespace to send and receive messages.  
+Namespace connection string can be defined via code using `ConnectionString` extension:
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: ~~via code~~

you show in a snippet that it's done via code, so this is redundant

> extension

did you mean "extension method"?

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+
+snippet: single_namespace_partitioning_strategy_with_default_connection_string
+
+or selecting explicitly the `SingleNamespacePartitioning` strategy and providing a map between namespace name and namespace connection string.
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > or selecting explicitly the `SingleNamespacePartitioning` strategy and providing a map between namespace name and namespace connection string.  

we're already discussing this strategy, so no need for explaining the first part, this will be enough:

or by providing a map between namespace name and namespace's connection string:

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+
+## Single namespace partitioning #
+
+The default _namespace partitioning strategy_ is `SingleNamespacePartitioning`. With this strategy, transport uses only a single namespace to send and receive messages.  
+Namespace connection string can be defined via code using `ConnectionString` extension:
+
+snippet: single_namespace_partitioning_strategy_with_default_connection_string
+
+or selecting explicitly the `SingleNamespacePartitioning` strategy and providing a map between namespace name and namespace connection string.  
+Transport uses namespace name to avoid sharing namespace secrets.
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > namespace secrets

that is very unofficial expression ;) how about:

In order to guarantee better security, the transport uses namespace name instead of raw connection strings:

or something similar

Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: That seems disconnected from the previous sentence. How about skipping it for now and linking when section on securing connection strings is ready? Something similar to http://docs.particular.net/nservicebus/azure-storage-queues/configuration#connection-strings-securing-connection-strings.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 33]
```diff
@@ -0,0 +1,63 @@
...
+
+snippet: single_namespace_partitioning_strategy_with_add_namespace
+
+With this strategy, only one namespace has to be configured. Configuration throws a `ConfigurationErrorsException` if more than one namespace has been configured.
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > With this strategy, only one namespace has to be configured. Configuration throws a `ConfigurationErrorsException` if more than one namespace has been configured.

Only one namespace can be configured when using `SingleNamespacePartitioning` strategy. When more than one namespace is specified, then NServiceBus throws a `ConfigurationErrorsException` at startup.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md (hunk 15 above)
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > relyable

reliable

> the easiest one 

what do you mean by that? easiest to configure? is that a default option?

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 34]
```diff
@@ -0,0 +1,63 @@
...
+
+With this strategy, only one namespace has to be configured. Configuration throws a `ConfigurationErrorsException` if more than one namespace has been configured.   
+`SingleNamespacePartitioning` strategy is the easiest one but it's the less relyable: the only configured namespace is a single point of failure for the entire system.
+
+## Round robin namespace partitioning #
+
+Selecting `RoundRobinNamespacePartitioning` multiple namespace are used. Messages are sent to a single namespace and received from all the namespace. For sending operations, namespaces are cycled through.  
+This strategy is useful to avoid throttling by Azure infrastructure for capacity limit.
+`RoundRobinNamespacePartitioning` strategy requires more than one configured namespace. Configuration throws a `ConfigurationErrorsException` if only one namespace has been configured. 
+
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: How about this:

The `RoundRobinNamespacePartitioning` strategy can be used in order to overcome capacity limits caused by Azure infrastructure throttling (link here to MSDN?). 

When using this strategy messages are sent to a single namespace, but can be sent from multiple namespaces.

Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > For sending operations, namespaces are cycled through.  

what does it? what is responsible for round robin?

Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: Multiple namespaces have to be configured when using SingleNamespacePartitioning strategy. When only one namespace is specified, then NServiceBus throws a `ConfigurationErrorsException` at startup.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 35]
```diff
@@ -0,0 +1,63 @@
...
+snippet: round_robin_partitioning_strategy
+
+## Fail over namespace partitioning #
+
+`FailOverNamespacePartitioning` strategy provides transport High Availability by using two different namespaces, a primary and a secondary. The transport uses the primary namespace. When the primary namespace is not available, the transport will switch to the secondary namespace.
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: Could you rephrase similar to https://github.com/Particular/docs.particular.net/pull/1425/files#r60874101?

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 36]
```diff
@@ -0,0 +1,63 @@
...
+`FailOverNamespacePartitioning` strategy requires exactly two namespaces. Configuration throws a `ConfigurationErrorsException` if number of namespaces is different than two.
+
+snippet: fail_over_partitioning_strategy 
+
+
+## Replicated namespace partitioning #
+
+Scenario covered by `RepliactedNamespacePartitioning` is to support publish messages across different data centers. With this strategy, transport uses all configured namespaces to send and receive messages.
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: Could you rephrase similar to https://github.com/Particular/docs.particular.net/pull/1425/files#r60874101?

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 37]
```diff
@@ -0,0 +1,63 @@
...
+`FailOverNamespacePartitioning` strategy provides transport High Availability by using two different namespaces, a primary and a secondary. The transport uses the primary namespace. When the primary namespace is not available, the transport will switch to the secondary namespace.
+`FailOverNamespacePartitioning` strategy requires exactly two namespaces. Configuration throws a `ConfigurationErrorsException` if number of namespaces is different than two.
+
+snippet: fail_over_partitioning_strategy 
+
+
+## Replicated namespace partitioning #
+
+Scenario covered by `RepliactedNamespacePartitioning` is to support publish messages across different data centers. With this strategy, transport uses all configured namespaces to send and receive messages.  
+When transport has to send a message, it sends the same message to all namespaces.
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: the same message or copy of that message to each namespace?

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 38]
```diff
@@ -0,0 +1,63 @@
...
+`RepliactedNamespacePartitioning` strategy requires more than one configured namespace. Configuration throws a `ConfigurationErrorsException` if only one namespace has been configured. 
+
+snippet: replicated_partitioning_strategy
+
+## Sharded namespace partitioning #
+
+With `ShardedNamespacePartitioning` messages are sent to the namespace selected by the custom `ShardingRule` defined and provided to the strategy.
```
Comment (weralabaj) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: To follow the convention for other strategies - when would you use it? What problem does it solve?

**Comment by weralabaj** [comment]
@melkio added a few comments

File: nservicebus/azure-service-bus/multiple-namespaces-support.md (hunk 15 above)
Comment (melkio) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > what do you mean by that? easiest to configure? is that a default option?

`SingleNamespacePartitioning` is the default option, and it's the easiest strategy implementation because it cover simple scenario: read/write from a single namespace. No HA or possibility to scale for instance

**Comment by melkio** [comment]
@weralabaj ping!

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 39]
```diff
@@ -0,0 +1,59 @@
...
+
+snippet: single_namespace_partitioning_strategy_with_add_namespace
+
+Only one namespace can be configured when using SingleNamespacePartitioning strategy. When more than one namespace is specified, then NServiceBus throws a `ConfigurationErrorsException` at startup.   
+`SingleNamespacePartitioning` strategy is the easiest one but it's the less reliable: the only configured namespace is a single point of failure for the entire system.
+
+## Round robin namespace partitioning
+
+Selecting `RoundRobinNamespacePartitioning` multiple namespace are used. Messages are sent to a single namespace and received from all the namespace. For sending operations, namespaces are cycled through.  
+This strategy is useful to avoid throttling by Azure infrastructure for [capacity limit](https://azure.microsoft.com/en-us/documentation/articles/service-bus-quotas/).
```
Comment (SeanFeldman) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: > his strategy is useful to avoid throttling by Azure infrastructure 

More like "This strategy is designed to avoid throttling by the service"
